app/services/sets/product_set.rb

```diff
@@ -114,6 +114,8 @@ def create_variant(product, variant_attributes)
 
       variant = product.variants.create(variant_attributes)
 
+      return false unless variant.save
```

Comment:
After seeing the other code, I would now suggest this:
```suggestion
      if variant.errors.present?
        product.errors.merge!(variant.errors)
        return false
      end
```

Then the ProductSet provides these error messages via its own `#errors` method and the old code in the controller should work.

Reply:
oh nice one! Will test it ;)